### PR TITLE
ENH: add Client.get_versions(check=False)

### LINKED ISSUE #664: ENH: add client.get_versions(check=True)
xref https://github.com/dask/distributed/pull/596

if check == True (default), then raise an error message showing differences in versions (e.g. distributed between workers and scheduler). makes debugging versions differences much easier.

open to different reporting options back to user (e.g. maybe just print out diffs)

### PR DESCRIPTION
**Pull request by jreback**
option to provide error checking .get_versions()

closes #664

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -1998,7 +2004,37 @@ def f(worker=None):
...
+            for pkg, cv in client_versions.items():
+                sv = scheduler_versions[pkg]
+                if sv != cv:
+                    raise ValueError("package [{package}] is version [{client}] "
+                                     "on client and [{scheduler}] on scheduler!".format(
+                                         package=pkg,
+                                         client=cv,
+                                         scheduler=sv))
+
+            for wkr, worker in workers.items():
```
Comment (mrocklin) [inline distributed/client.py]: We often use the pronouns `w` or `worker` to mean a worker address, like `127.0.0.1:8786`.  Recommend changing `wkr` to w or worker and changing worker to some other pronoun like `response` or `d`

File: distributed/client.py [hunk 2]
```diff
@@ -1998,7 +2004,37 @@ def f(worker=None):
                 return None
 
         workers = sync(self.loop, self._run, f)
-        return {'scheduler': scheduler, 'workers': workers, 'client': client}
+        result = {'scheduler': scheduler, 'workers': workers, 'client': client}
+
+        if check:
+
```
Comment (mrocklin) [inline distributed/client.py]: Remove excess whitespace

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -3281,6 +3281,10 @@ def test_get_versions(loop):
             for k, v in v['workers'].items():
                 assert v is not None
 
+            # smoke test for versions
+            # that this does not raise
+            c.get_versions(check=True)
```
Comment (mrocklin) [inline distributed/tests/test_client.py]: Would prefer these comments after the code rather than on top.

**Comment by jreback** [comment]
updated.

note that this is *not* extensively tested......

**Comment by mrocklin** [comment]
It's hard to test things that explicitly need different software environments

**Comment by jreback** [comment]
can add to release-notes.md if you'd like (ideally you'd add a new section for the next release when releasing FYI)

**Comment by mrocklin** [comment]
Yeah, we're not very disciplined about asking contributors to add to release-notes.  I tend to just do things manually at each release.  However adding something yourself would be very appreciated.

**Comment by jreback** [comment]
http://dask.pydata.org/en/latest/develop.html is slightly outdated I think. should be ``realease-notes.md``? (or is this different for dask & distributed)?

**Comment by mrocklin** [comment]
Different between the two unfortunately.

**Comment by jreback** [comment]
updated

**Comment by jreback** [comment]
would it be useful to instead of a ``check=False`` have

``check_required``, ``check_optional`` and ``check`` (which would override the others if True)?

**Pull request merged**